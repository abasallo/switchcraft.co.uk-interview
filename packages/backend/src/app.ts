import express from "express";
import { appLogger } from "./app-logger";
import { handleApiErrors } from "./handle-error";
import { getBigGreenEnergyDeals } from "./big-green-energy-deals/get-big-green-energy-deals";
import type { FrontendDealsResponse } from "@switchcraft-interview/shared-models";
import cors from "cors";

const app = express();
app.use(cors());
app.use(express.json());

app.use((req, res, next) => {
  appLogger.trace({
    url: req.url,
    body: req.body as {},
    params: req.params,
    query: req.query
  }, "Got request");
  next();
});

app.get<undefined, FrontendDealsResponse>("/deals", async (req, res) => {
  const deals = await getBigGreenEnergyDeals();
  return res.json({
    deals
  });
});

app.use(handleApiErrors);

export { app };
