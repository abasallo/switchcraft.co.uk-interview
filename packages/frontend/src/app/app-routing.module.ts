import { NgModule } from "@angular/core";
import { Routes } from "@angular/router";
import { RouterModule } from "@angular/router";
import { AllDealsComponent } from "./all-deals/all-deals.component";

const routes: Routes = [
  {
    path: "",
    component: AllDealsComponent
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
