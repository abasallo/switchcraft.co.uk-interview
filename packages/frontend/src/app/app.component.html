<div class="nav"></div>
<router-outlet></router-outlet>
