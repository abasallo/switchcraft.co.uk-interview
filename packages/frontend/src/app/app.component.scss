.nav {
  height: 94px;
  background-color: #111155;
  margin-bottom: 34px;
}
