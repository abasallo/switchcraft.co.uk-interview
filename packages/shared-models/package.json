{
  "name": "@switchcraft-interview/shared-models",
  "version": "1.0.0",
  "main": "./dist/src/index.js",
  "license": "MIT",
  "dependencies": {
    "typescript": "^4.2.4"
  },
  "scripts": {
    "build": "yarn tsc",
    "prepublish": "yarn build"
  }
}
