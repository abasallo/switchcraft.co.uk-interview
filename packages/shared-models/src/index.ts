export * from "./big-green-energy-deal.interface";

export * from "./shiny-yellow-energy-deal.interface";

export * from "./frontend-deal.interface";

export * from "./frontend-deals-response.interface";
