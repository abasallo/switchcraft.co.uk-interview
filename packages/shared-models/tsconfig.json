{
  "extends": "../../tsconfig.json",
  "include": [
    "./src/*"
  ],
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./",
    "baseUrl": "./src",
    "declaration": true
  }
}
