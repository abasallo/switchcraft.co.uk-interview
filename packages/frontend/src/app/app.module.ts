import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";
import { AppRoutingModule } from "./app-routing.module";
import { AppComponent } from "./app.component";
import { SingleDealComponent } from "./single-deal/single-deal.component";
import { AllDealsComponent } from "./all-deals/all-deals.component";
import { HttpClientModule } from "@angular/common/http";

@NgModule({
  declarations: [
    AppComponent,
    SingleDealComponent,
    AllDealsComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    HttpClientModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
