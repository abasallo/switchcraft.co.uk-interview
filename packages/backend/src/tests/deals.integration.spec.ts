import request from "supertest";
import express from "express";
import { app } from "../app";
import { FrontendDealsResponse } from "@switchcraft-interview/shared-models";
import { MongoClient } from "mongodb";

const wrapper = express();
wrapper.use("/", app);

const dealsUrl = "/deals";

describe("deals endpoint", () => {
  let result: FrontendDealsResponse;
  let connection: MongoClient;

  const initDatabase = async (): Promise<void> => {
    connection = await MongoClient.connect(
      process.env.MONGO_URL!,
      {
        useNewUrlParser: true
      }
    );
    const db = await connection.db("interview");
    await db.collection("bigGreenEnergyDeals").insertMany([
      {
        name: "Fixed Green Deal",
        supplierName: "Big Green Energy",
        annualCost: 450,
        isGreenEnergy: true,
        durationMonths: 12,
        logo: "<svg></svg>"
      },
      {
        name: "Big Green Deal",
        supplierName: "Big Green Energy",
        annualCost: 500,
        isGreenEnergy: true,
        logo: "<svg></svg>"
      }
    ]);
    await db.collection("shinyYellowEnergyDeals").insertMany([
      {
        name: "Shiny Forever",
        monthlyCost: 50,
        green: false,
        logo: "<svg></svg>"
      },
      {
        name: "Shiny 5000",
        monthlyCost: 30,
        green: true,
        duration: "18",
        logo: "<svg></svg>"
      }
    ]);
  };

  beforeAll(async () => {
    await initDatabase();
    const { body } = await request(wrapper).get(dealsUrl);
    result = body;
  });

  afterAll(async () => {
    await connection.close();
  });

  test("returns the expected result", () => {
    expect(result).toMatchSnapshot();
  });

});
