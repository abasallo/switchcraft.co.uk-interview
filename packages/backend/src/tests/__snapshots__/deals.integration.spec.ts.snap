// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deals endpoint returns the expected result 1`] = `
Object {
  "deals": Array [
    Object {
      "dealName": "Fixed Green Deal",
      "durationMonths": 12,
      "isGreenEnergy": true,
      "logo": "<svg></svg>",
      "supplierName": "Big Green Energy",
      "yearlyCost": 450,
    },
    Object {
      "dealName": "Big Green Deal",
      "isGreenEnergy": true,
      "logo": "<svg></svg>",
      "supplierName": "Big Green Energy",
      "yearlyCost": 500,
    },
  ],
}
`;
